lib/msal-common/src/client/SilentFlowClient.ts

```diff
@@ -0,0 +1,161 @@
+import { TimeUtils } from "../utils/TimeUtils";
+import { RefreshTokenRequest } from "../request/RefreshTokenRequest";
+import { RefreshTokenClient } from "./RefreshTokenClient";
+import { ClientAuthError } from "../error/ClientAuthError";
+import { CredentialFilter, CredentialCache } from "../unifiedCache/utils/CacheTypes";
+
+export class SilentFlowClient extends BaseClient {
+
+    constructor(configuration: ClientConfiguration) {
+        super(configuration);
+    }
+
+    /**
+     * Retrieves a token from cache if it is still valid, or uses the cached refresh token to renew
+     * the given token and returns the renewed token
+     * @param request
+     */
+    public async acquireToken(request: SilentFlowRequest): Promise<AuthenticationResult>{
+        let cacheRecord: CacheRecord;
+        let idTokenObj: IdToken;
+        const requestScopes = new ScopeSet(request.scopes || [], this.config.authOptions.clientId, true);
+
+        // We currently do not support silent flow for account === null use cases; This will be revisited for confidential flow usecases
+        if (request.account === null) {
+            throw ClientAuthError.createNoAccountInSilentRequestError();
+        } else {
+            // fetch account
+            const accountKey: string = CacheHelper.generateAccountCacheKey(request.account);
+            cacheRecord.account = this.unifiedCacheManager.getAccount(accountKey);
+
+            const homeAccountId = cacheRecord.account.homeAccountId;
+            const environment = cacheRecord.account.environment;
+
+            // fetch idToken, accessToken, refreshToken
+            cacheRecord.idToken = this.fetchIdToken(homeAccountId, environment, cacheRecord.account.realm);
+            idTokenObj = new IdToken(cacheRecord.idToken.secret, this.config.cryptoInterface);
+            cacheRecord.accessToken = this.fetchAccessToken(homeAccountId, environment, requestScopes, cacheRecord.account.realm);
+            cacheRecord.refreshToken = this.fetchRefreshToken(homeAccountId, environment);
+
+            // If accessToken has expired, call refreshToken flow to fetch a new set of tokens
+            if (request.forceRefresh || (!!cacheRecord.accessToken && this.isTokenExpired(cacheRecord.accessToken.expiresOn))) {
+                // check if we have refreshToken
+                if (!!cacheRecord.refreshToken) {
+                    const refreshTokenClient = new RefreshTokenClient(this.config);
+                    const refreshTokenRequest: RefreshTokenRequest = {
+                        scopes: request.scopes,
+                        refreshToken: cacheRecord.refreshToken.secret,
+                        authority: request.authority
+                    };
+
+                    return refreshTokenClient.acquireToken(refreshTokenRequest);
+                }
+                // no refresh Token
+                else {
+                    throw ClientAuthError.createNoTokenInCacheError();
```

Comment:
I think this should be a unique error relating to not having a valid refresh token.

Reply:
Also, can you move checks for error conditions to the top? That way you don't need `else`.

Reply:
done.